Source/ExecutionEngine/WorkStealingTaskScheduler.cs

```diff
@@ -0,0 +1,325 @@
+// Code taken from https://github.com/ChadBurggraf/parallel-extensions-extras/blob/ec803e58eee28c698e44f55f49c5ad6671b1aa58/TaskSchedulers/WorkStealingTaskScheduler.cs
```

Comment:
It is, but I handrolled something else now.

Reply:
Ah! I just started looking at the comment trail on this PR and it took a while for things to make sense. It sounds like @keyboardDrummer was using a copied version of some publicly-available code but has now changed the approach to a simple threadpool class with a large stack size. I saw the new class and indeed it looks quite simple.

Is there any reason to gate the use of the new threadpool under a command-line flag? In other words, do you think it is possible for the new threadpool to be worse than the custom threadpool in some way. If so, might be better to use a command-line option, especially since the solution is temporary anyway.

Reply:
> Is there any reason to gate the use of the new threadpool under a command-line flag? In other words, do you think it is possible for the new threadpool to be worse than the custom threadpool in some way. If so, might be better to use a command-line option, especially since the solution is temporary anyway.

I think the main downside of this PR is that the thread pool it introduces uses some memory, 16MB * the number of processors. I think any degradation in the running time will tiny in the scheme of doing verification using Boogie.

Note that older Boogies, like 2.9.4, also spawned 16MB stack threads, as you can see here: https://github.com/boogie-org/boogie/blob/v2.9.4/Source/ExecutionEngine/ExecutionEngine.cs#L433. They spawned one for each Boogie implementation: https://github.com/boogie-org/boogie/blob/v2.9.4/Source/ExecutionEngine/ExecutionEngine.cs#L1018
but those could be garbage collected after an implementation had finished running, so the memory usage was limited to 16MB * implementations_running_concurrently, which at most is the same as what this PR does. Note that this PR can re-use threads between implementations since it uses a pool.

I don't want to use the single-use threads `TaskScheduler` of 2.9.4, since the current Boogie makes much heavier use of async/await so it's important for the `TaskScheduler` to be able to queue `Task`s cheaply.

Reply:
Please compare the running time of the current Boogie and your new version on a few large-ish examples, potentially generated from Dafny. Another option could be to compare the running time for the entire Boogie regression testsuite. Just a sanity check. I expect your intuition that there wouldn't be any significant difference to be valid.